[Tablet Orders] Variation removal from Order Summary sync to open Variation List.

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #11932
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

When a variable product is removed from the order using the button in the order summary, we need to show that accurately in the variation list as well, to avoid confusing the merchant.

When the variation list was open, previously we did not update the selection effectively – it would only be updated if the list was closed and reopened.

This PR ensures the selections are kept in sync even when the list is open, by accessing the active variation list view model and removing the selection there.

In this PR, we:
1. Move the storage of the state for whether/which variation list is open from the ProductSelectorView, to its view model.
2. Call `removeSelection` with the appropriate variation ID when a user taps `Remove product from order` in the order summary.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

1. Enable the `sideBySideViewForOrderForm` feature flag
2. Launch the app on an iPad or large iPhone in landscape
3. Navigate to `Orders`, and tap `+`
4. Tap on a variable product
5. Add a variation to your order – observe that it shows as selected in the list
6. Expand the variation in the OrderSummary
7. Tap `Remove product from order`
8. Observe that the variation is no longer selected in the list

### Test coupons:
1. Enable the `sideBySideViewForOrderForm` feature flag
2. Launch the app and navigate to `Menu > Coupons`
3. Add a coupon
4. Select Product coupon
5. Tap `All Products`
6. Tap a checkbox for a variable product
7. Observe that all the variations are selected
8. Tap the row for the variable product
9. Change how many are selected
10. Tap back
11. Observe that the variation selection state is `-`, and the correct number of products is mentioned in the button.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/woocommerce/woocommerce-ios/assets/2472348/b23da6bf-466f-4846-ad5d-529c98b07545



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

joshheald committed:
11932 Move variation list logic to view model

joshheald committed:
11932 update variation list with removals from order

When a variable product is removed from the order using the button in the order summary, we need to show that accurately in the variation list as well, to avoid confusing the merchant.

When the variation list was open, previously we did not update the selection effectively – it would only be updated if the list was closed and reopened.

This commit ensures the selections are kept in sync even when the list is open, by accessing the active variation list view model and removing the selection there.

joshheald committed:
Fix whitespace

joshheald committed:
11932 Ensure variation list shown for the right ID

joshheald committed:
11932 Fix selection update on back tap

Moving the `variationListViewModel` to the `ProductSelectorViewModel` (rather than a state variable on the `ProductSelectorView`) resulted in a lifecycle issue where tapping back would not store the updates.

`onDisappear` wasn’t a great place to put this call anyway, it would be called for other changes which aren’t related to the user tapping back.

Since we already add the back button ourselves, and disable swiping back, this is a straightforward fix for non-immediate sync contexts

iamgabrielma commented:
We have a regression regarding variation selection when we have the feature flag enabled, but we do not load a side-to-side view (eg: iPhone portrait):
1. Turn the flag to true, run the app on iPhone portrait
1. Add variations to order
2. Remove them by tapping on the collapsible product card for each product, then the "Remove product from order" button
3. Observe that when we open the product selector again, these remain selected

![Simulator Screen Recording - iPhone 15 - 2024-02-27 at 10 40 38](https://github.com/woocommerce/woocommerce-ios/assets/3812076/eb131b9b-5167-402f-a1f0-01edb1547569)

I think this could be related to the new logic for updating selections only working when we `syncChangesImmediately`, which doesn't apply in iPhone portrait?

iamgabrielma commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/ProductSelectorViewModelTests.swift:
```diff
@@ -538,33 +538,72 @@ final class ProductSelectorViewModelTests: XCTestCase {
+        // When
+        viewModel.variationCheckboxTapped(for: product.productID)
+
+        // Then
+        let variationsViewModel = viewModel.productVariationListViewModel
+        let actualViewModel = try XCTUnwrap(variationsViewModel)
+        XCTAssertEqual(actualViewModel.productName, product.name)
+        XCTAssertTrue(viewModel.isShowingProductVariationList)
```
WDYT if we refactor the `actualViewModel` here? Does it feels easier to read?
```suggestion
        // Then
        let variationsViewModel = try XCTUnwrap(viewModel.productVariationListViewModel)
        XCTAssertEqual(variationsViewModel.productName, product.name)
        XCTAssertTrue(viewModel.isShowingProductVariationList)
```

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductVariationSelectorView.swift:
```diff
@@ -92,6 +92,7 @@ struct ProductVariationSelectorView: View {
             // Minimal back button
             ToolbarItem(placement: .navigationBarLeading) {
                 Button {
+                    onMultipleSelections?(viewModel.selectedProductVariationIDs)
```
Small but great change 💯

joshheald commented:
> We have a regression regarding variation selection when we have the feature flag enabled, but we do not load a side-to-side view (eg: iPhone portrait):
> ...
> I think this could be related to the new logic for updating selections only working when we `syncChangesImmediately`, which doesn't apply in iPhone portrait?

Thanks, good catch! I'll take a look at this later today/tomorrow, sorry for missing it.

joshheald committed:
11932 Always update selections on product selector

Previously, the product selector could get out of sync after products were removed from the Order Form’s product rows.

This commit ensures we always update the productSelector to match the OrderForm when changes are made there.

joshheald commented:
> > We have a regression regarding variation selection when we have the feature flag enabled, but we do not load a side-to-side view (eg: iPhone portrait):
> > ...
> > I think this could be related to the new logic for updating selections only working when we `syncChangesImmediately`, which doesn't apply in iPhone portrait?
> 
> Thanks, good catch! I'll take a look at this later today/tomorrow, sorry for missing it.

@iamgabrielma fixed in c9294bd

joshheald committed:
11932 minor test style change

iamgabrielma reviewed (approved):
Works great! :shipit:
